### PR TITLE
8334239: Introduce macro for ubsan method/function exclusions

### PR DESCRIPTION
**Pull request by MBaesken**
A number of functions/methods have to be excluded from ubsan detection (e.g. because they do things that ubsan warns about, however it is still valid what is done there).
We can simplify this by introducing a macro (similar to asan-related ATTRIBUTE_NO_ASAN, see sanitizers/address.hpp).
Currently something like this is used :
```
#if defined(__clang__) || defined(__GNUC__)
__attribute__((no_sanitize("undefined")))
#endif
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334239](https://bugs.openjdk.org/browse/JDK-8334239): Introduce macro for ubsan method/function exclusions (**Bug** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [c07f9324](https://git.openjdk.org/jdk/pull/19722/files/c07f9324c81246cf4149c49d48a3af729ea2acee)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19722/head:pull/19722` \
`$ git checkout pull/19722`

Update a local copy of the PR: \
`$ git checkout pull/19722` \
`$ git pull https://git.openjdk.org/jdk.git pull/19722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19722`

View PR using the GUI difftool: \
`$ git pr show -t 19722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19722.diff">https://git.openjdk.org/jdk/pull/19722.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19722#issuecomment-2168087564)

### CODE REVIEW

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
In keeping with address.hpp and leak.hpp the header would either be called undefinedbehaviour.hpp or just ub.hpp for short.

I think the actual attribute would read better if placed on the line before the function definition.

Thanks.

**Comment by tstuefe** [comment]
> In keeping with address.hpp and leak.hpp the header would either be called undefinedbehaviour.hpp or just ub.hpp for short.
> 
> I think the actual attribute would read better if placed on the line before the function definition.
> 
> Thanks.

I vote for ubsan.hpp and asan.hpp, or for a common header sanitizer.hpp.

**Comment by MBaesken** [comment]
> I think the actual attribute would read better if placed on the line before the function definition.

I placed it on the line before the function.

**Comment by dholmes-ora** [comment]
> I vote for ubsan.hpp and asan.hpp, or for a common header sanitizer.hpp.

That's a suggestion that should have been made when we added asan and lsan support at the beginning if last year. The "san" in the name is redundant with the "sanitizers" in the path.

**Comment by tstuefe** [comment]
> Under 'sanitizers' there would be a header for one sanitizer called 'ubsan' - that naming might be a little redundant but would still make sense because that#s the name of the tool ; but I can live with both ub.hpp or ubsan.hpp . Are there any other opinions ?

I'm fine with all proposals so far.

**Comment by MBaesken** [comment]
I changed the header name from ubsan.hpp to ub.hpp .

File: src/hotspot/share/sanitizers/ub.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#ifndef SHARE_SANITIZERS_UB_HPP
```
Comment (stefank) [inline src/hotspot/share/sanitizers/ub.hpp:25]:
```suggestion
 */

#ifndef SHARE_SANITIZERS_UB_HPP
```

File: src/hotspot/share/sanitizers/ub.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+// example causing desired signals/crashes or overflows
+#if defined(__clang__) || defined(__GNUC__)
+#define ATTRIBUTE_NO_UBSAN __attribute__((no_sanitize("undefined")))
+#endif
+
+#ifndef ATTRIBUTE_NO_UBSAN
+#define ATTRIBUTE_NO_UBSAN
+#endif
+
+
```
Comment (stefank) [inline src/hotspot/share/sanitizers/ub.hpp:42]:
```suggestion

```

File: src/hotspot/share/sanitizers/ub.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ */
+#ifndef SHARE_SANITIZERS_UB_HPP
+#define SHARE_SANITIZERS_UB_HPP
+
+// ATTRIBUTE_NO_UBSAN
+//
+// Function attribute which informs the compiler to disable UBSan checks in the
+// following function or method.
+// Useful if the function or method is known to do something special or even 'dangerous', for
+// example causing desired signals/crashes or overflows
```
Comment (stefank) [inline src/hotspot/share/sanitizers/ub.hpp:33]:
```suggestion
// example causing desired signals/crashes or overflows.
```

**Review by stefank — COMMENTED** [review]
I've listed some whitespace and punctuation nits.

I have a question/comment about this comment:
```
... special or even 'dangerous', for example causing desired signals/crashes or *overflows*.
```

Do we really intend to use ATTRIBUTE_NO_UBSAN to silence overflow issues? Don't we want to fix all of those?

**Review by tstuefe — APPROVED** [review]
good, if you take @stefank's suggestions

**Comment by MBaesken** [comment]
> Do we really intend to use ATTRIBUTE_NO_UBSAN to silence overflow issues? Don't we want to fix all of those?

Some stuff needs fixing in external libs, that can take some  (maybe a lot) of  time.  Other coding might already handle overflows but only after  the overflow happened; so ubsan is triggered  but the coding is fine.
For now I would be in a first step already happy to be able to *build* with ubsan enabled  (we are close to this but not 100% there, I still have to use some patches on top of OpenJDK).

**Comment by stefank** [comment]
> > Do we really intend to use ATTRIBUTE_NO_UBSAN to silence overflow issues? Don't we want to fix all of those?
> 
> Some stuff needs fixing in external libs, that can take some (maybe a lot) of time.

Will ATTRIBUTE_NO_UBSAN be used in those case?

> Other coding might already handle overflows but only after the overflow happened; so ubsan is triggered but the coding is fine.

This is the part that I'm questioning. I don't think this is true.

**Comment by MBaesken** [comment]
I can remove the 2 word 'or overflows' if you want .
(there was at least a 'division by 0' example case here https://github.com/openjdk/jdk/pull/19674
where the case detected by ubsan was already handled; but the coding could be reorganized to avoid it at all ;
comment in that review from the area expert 'This whole ubsan thing is just random findings based on what tests we have.
We'll never be ubsan clean and I'm not sure it is advisable in all cases.' )

**Comment by stefank** [comment]
> I can remove the 2 word 'or overflows' if you want.

Yes, please. Maybe there are places where overflows are OK, but I would like to get that better explained before we hint people that they can use ATTRIBUTE_NO_UBSAN to "silence" ubsan.

> the case detected by ubsan was already handled

I've seen this argument a number of times in the UBSAN PRs. My concern is that I'm not convinced that you can "handle" real UB issue. You can work around or fix them, but if you have a real UB and and try to silence ubsan, then you still have the UB and the compiler is within rights to throw away the code that "handles" the issue. Over the years we have seen a few issues where this happens: some if statements check for overflows, but the checks were unexpectedly removed by the compiler.

File: src/hotspot/share/sanitizers/ub.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// Useful if the function or method is known to do something special or even 'dangerous', for
+// example causing desired signals/crashes.
+#if defined(__clang__) || defined(__GNUC__)
+#define ATTRIBUTE_NO_UBSAN __attribute__((no_sanitize("undefined")))
+#endif
+
+#ifndef ATTRIBUTE_NO_UBSAN
+#define ATTRIBUTE_NO_UBSAN
+#endif
+#endif // SHARE_SANITIZERS_UB_HPP
```
Comment (stefank) [inline src/hotspot/share/sanitizers/ub.hpp:42]: You removed one too many blank lines here. It used to be two, now there are none.

**Review by stefank — COMMENTED** [review]
Thanks for making the change. I think this looks good, minus one nit in the whitespace change.

**Review by stefank — APPROVED** [review]

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by MBaesken** [comment]
Thanks for the reviews !

/integrate

**Review by dholmes-ora — COMMENTED** [review]
Apologies - my review of your updates in response to my comments never got submitted.

**Pull request closed without merging**